allennlp/data/semparse/worlds/nlvr_world.py

```diff
@@ -0,0 +1,509 @@
+        executable as an attribute function, and the last element should be a constant.
+        TODO(pradeep): We may want to change the order of arguments here to make decoding easier.
+
+        Syntax: assertion_function(attribute_function, constant)
+        """
+        assert isinstance(sub_expression, list), "Invalid assertion expression: %s" % sub_expression
+        if len(sub_expression) == 1 and isinstance(sub_expression[0], list):
+            return self._execute_assertion(sub_expression[0])
+        assert isinstance(sub_expression[0], str) and sub_expression[0].startswith("assert_"),\
+               "Invalid assertion function: %s" % (sub_expression[0])
+        function = getattr(self, sub_expression[0])
+        arguments = sub_expression[1]
+        target_attribute = self._execute_constant(arguments[2])
+        if arguments[0] == "count":
+            counted_value = self._execute_count_function(arguments[:2])
+            return function(counted_value, target_attribute)
+        attribute_set = self._execute_attribute_function(arguments[:2])
+        # We defined attribute functions to always return sets. But when this assertion is being executed, for the
+        # assertion to be true, the returned set needs to be singleton. If not, the assertion should fail.
+        if len(attribute_set) != 1:
+            return False
+        (attribute,) = attribute_set
+        return function(attribute, target_attribute)
+
+    def _execute_attribute_function(self, sub_expression: List) -> Set[str]:
+        """
+        Attribute functions return a set of evaluated attributes. The function has to be ``color``
+        or ``shape``, and the only element in the nested list should evaluate to a set of objects.
+
+        Syntax: attribute_function(input_set)
+        """
+        assert isinstance(sub_expression, list), "Invalid attribute expression: %s" % sub_expression
+        function_name = sub_expression[0]
+        arguments_list = sub_expression[1]
+        input_set = self._execute_object_filter(arguments_list)
+        if function_name == "shape":
+            return self.shape(input_set)
+        elif function_name == "color":
+            return self.color(input_set)
+        else:
+            raise RuntimeError("Invalid attribute function: %s" % sub_expression[0])
+
+    def _execute_count_function(self, sub_expression: List) -> int:
+        """
+        Acceptable syntax is count(object_set).
+        """
+        arguments_list = sub_expression[1]
+        if arguments_list[0].startswith("filter_") or arguments_list[0] == "all_boxes":
+            return self.count(self._execute_box_filter(arguments_list))
+        else:
+            return self.count(self._execute_object_filter(arguments_list))
+
+    def _execute_box_filter(self, sub_expression: Union[str, List]) -> Set[Box]:
+        """
+        Box filtering functions either apply a filter on a set of boxes and return the filtered set.
```

Comment:
"either"?

Reply:
I didn't finish the sentence. I meant to say that they either apply a filter or return all boxes. Fixed this.